### PR TITLE
[PW-2578] Add ability to store mapping of metabase card ids

### PR DESCRIPTION
**Pull request by prasanthkrajan**
[Add ability to store mapping of metabase card ids](https://perxtechnologies.atlassian.net/browse/PW-2578)

# Describe the changes made. What does this PR changes that might be critical. If any critical decisions have been made, make sure you explain the rationale for these decisions.

### Model
- Added a MetabaseCard model that holds card_id (integer) and identifier(string) UNIQUE, which maps to the card_id when a unique identifier. This is meant to be created by the data team.

### Endpoint
- GET /metabase_token/<identifier | id> to retrieve metabase token. The support for id is deprecated and once the FE is fully using the identifier feature will be disabled.
- CRUD metabase_card to create a card identifier record

# How does the implementation addresses the problem

This allows front end to pass an identifier or an ID to retrieve the metabase token.

### CODE REVIEW

File: services/cognito/app/controllers/metabase_token_controller.rb [hunk 1]
```diff
@@ -29,4 +31,18 @@ def payload_params
...
+
+    if card_map.card_identifier_record.present?
+      params[:id] = card_map.mapped_value
+    else
+      render json: { errors: 'Card ID not found' }
+    end
+  end
+
+  def card_map
+    Metabase::CardMap.new(identifier: params[:identifier])
```
Comment (rpbaltazar) [inline services/cognito/app/controllers/metabase_token_controller.rb]: Creating this seems like a slight overkill. In the end we just need the id for the card. its a simple find_by

File: services/cognito/app/services/metabase/card_map.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+module Metabase
```
Comment (rpbaltazar) [inline services/cognito/app/services/metabase/card_map.rb]: definitely overkill for searching for the card id

File: services/cognito/config/routes.rb [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
+  jsonapi_resources :metabase_cards
   jsonapi_resources :chown_results
   jsonapi_resources :chown_requests, only: [:create]
   jsonapi_resources :identifiers
   jsonapi_resources :users
   jsonapi_resources :pools
   jsonapi_resources :endpoints
 
-  resources :metabase_token, only: :show
+  get '/metabase_token/:identifier' => 'metabase_token#show'
```
Comment (rpbaltazar) [inline services/cognito/config/routes.rb]: Point this to its own controller method

File: services/cognito/app/controllers/metabase_token_controller.rb [hunk 4]
```diff
@@ -29,4 +31,18 @@ def payload_params
 
     options.merge(params[:params].to_unsafe_h.deep_symbolize_keys)
   end
+
+  def map_identifier_to_id
```
Comment (rpbaltazar) [inline services/cognito/app/controllers/metabase_token_controller.rb]: create the custom controller method and do the find here:

```
def show_identifier
    identifier = params.delete(:identifier)
    card = MetabaseCard.find_by(uniq_identifier: identifier)
    render json: { errors: 'Card ID not found' } if card.nil?

    params[:id] = card.card_id
    show
end
```

File: services/cognito/spec/requests/metabase_token_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
+require 'rails_helper'
```
Comment (rpbaltazar) [inline services/cognito/spec/requests/metabase_token_spec.rb]: rubocop will complain here

File: services/cognito/spec/requests/metabase_token_spec.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+          get url, headers: request_headers
+        end
+
+        it 'returns an error' do
+          expect(errors.size).to be_positive
+        end
+      end
+    end
+  end
+end
```
Comment (rpbaltazar) [inline services/cognito/spec/requests/metabase_token_spec.rb]: and here

File: services/cognito/spec/factories/metabase_cards.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
```
Comment (rpbaltazar) [inline services/cognito/spec/factories/metabase_cards.rb]: rubocop will complain here

Comment (prasanthkrajan) [reply services/cognito/spec/factories/metabase_cards.rb]: @rpbaltazar pushed requested changes. ready for review.

File: services/cognito/spec/requests/metabase_cards_spec.rb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      include_examples 'unauthenticated get'
+    end
+
+    context 'Authenticated user' do
+      include_context 'authorized user'
+
+      let(:model_data) { build(:metabase_card) }
+
+      before do
+        mock_authentication if mock
```
Comment (rpbaltazar) [inline services/cognito/spec/requests/metabase_cards_spec.rb]: No longer needed

File: services/cognito/spec/models/metabase_card_spec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  let(:record_one)   { build(factory_name, card_id: rand(1..20), identifier: subject.identifier) }
+  let(:record_two)   { build(factory_name, card_id: nil) }
+  let(:record_three) { build(factory_name, identifier: nil) }
+  let(:record_four)  { build(factory_name, card_id: subject.card_id) }
+
+  it 'is valid if card_id and identifier present' do
+    expect(subject).to be_valid
+  end
+
+  it 'is invalid if uniq_identifer is not unique' do
```
Comment (rpbaltazar) [inline services/cognito/spec/models/metabase_card_spec.rb]: I've renamed `uniq_identifier` to identifier FYI

File: services/cognito/app/controllers/metabase_token_controller.rb [hunk 10]
```diff
@@ -12,6 +12,18 @@ def show
     end
   end
 
+  def show_identifier
+    identifier = params.delete(:identifier)
```
Comment (rpbaltazar) [inline services/cognito/app/controllers/metabase_token_controller.rb]: Lets discuss this tomorrow. the idea is that we should support both id and identifier, so at this point we might not even need the custom endpoint.

**Pull request merged**